torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -907,6 +908,30 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
   return true;
 }
 
+ProcessGroupNCCL::DebugInfoWriter::DebugInfoWriter(int rank) {
+  const char* fileName = parseEnvVarString(
+      "TORCH_NCCL_DEBUG_INFO_TEMP_FILE", "/tmp/nccl_trace_rank_");
+  filename_ = c10::str(std::string(fileName), rank);
+}
+
+ProcessGroupNCCL::DebugInfoWriter::~DebugInfoWriter() = default;
+
+void ProcessGroupNCCL::DebugInfoWriter::write(const std::string& ncclTrace) {
+  // Open a file for writing. The ios::binary flag is used to write data as
+  // binary.
+  std::ofstream file(filename_, std::ios::binary);
+
+  // Check if the file was opened successfully.
+  if (!file.is_open()) {
+    LOG(ERROR) << "Error opening file for writing NCCLPG debug info: "
+               << filename_;
+    return;
+  }
+
+  file.write(ncclTrace.data(), ncclTrace.size());
```

Comment:
Is this write behavior on by default or gated by an env? We should not directly write logs into user's disk by default.

Reply:
It's gated and default behavior is for OSS users, advanced users will override the write behavior.

Reply:
Sorry which env gates it? I might have missed it.

Reply:
If users set `TORCH_NCCL_TRACE_BUFFER_SIZE` larger than 0, users will enable the flight recorder. `TORCH_NCCL_ENABLE_MONITORING` will enable the dumping and monitor thread.

Reply:
Just a thought, not necessary for this PR: we can rely on whether `TORCH_NCCL_DEBUG_FILE` is set to enable/disable writing.

Reply:
yeah, we can do that. But if you look at the implementation of the default write, we don't write the the file if `!file.is_open()`.